lib/rubocop/cop/commissioner.rb

```diff
@@ -94,6 +94,12 @@ def investigate(processed_source)
 
       private
 
+      def begin_investigation(processed_source, offset:, original:)
+        @cops.each do |cop|
+          cop.send(:begin_investigation, processed_source, offset: offset, original: original)
```

Comment:
`send` is slow in general. Can you move the `begin_investigation` method from private and mark `# @api private`?

```suggestion
          cop.begin_investigation(processed_source, offset: offset, original: original)
```

This is the difference for reference.

```diff
diff --git a/lib/rubocop/cop/base.rb b/lib/rubocop/cop/base.rb
index 3dc069cee..fc51bc9ff 100644
--- a/lib/rubocop/cop/base.rb
+++ b/lib/rubocop/cop/base.rb
@@ -287,6 +287,21 @@ module RuboCop
         end
       end

+      # Called before any investigation.
+      # @api private
+      def begin_investigation(processed_source, offset: 0, original: processed_source)
+        @current_offenses = []
+        @current_offense_locations = nil
+        @currently_disabled_lines = nil
+        @processed_source = processed_source
+        @current_corrector = Corrector.new(@processed_source) if @processed_source.valid_syntax?
+
+        # We need to keep track of the original source and offset,
+        # because `processed_source` here may be an embedded code in it.
+        @current_offset = offset
+        @current_original = original
+      end
+
       private

       ### Reserved for Cop::Cop
@@ -313,21 +328,6 @@ module RuboCop
         @restrict_on_send ||= self::RESTRICT_ON_SEND.to_a.freeze
       end

-      # Called before any investigation
-      def begin_investigation(processed_source, offset: 0,
-                              original: processed_source)
-        @current_offenses = []
-        @current_offense_locations = nil
-        @currently_disabled_lines = nil
-        @processed_source = processed_source
-        @current_corrector = Corrector.new(@processed_source) if @processed_source.valid_syntax?
-
-        # We need to keep track of the original source and offset,
-        # because `processed_source` here may be an embedded code in it.
-        @current_offset = offset
-        @current_original = original
-      end
-
       # Called to complete an investigation
       def complete_investigation
         InvestigationReport.new(self, processed_source, @current_offenses, @current_corrector)
diff --git a/lib/rubocop/cop/commissioner.rb b/lib/rubocop/cop/commissioner.rb
index 85820dea2..412db6f26 100644
--- a/lib/rubocop/cop/commissioner.rb
+++ b/lib/rubocop/cop/commissioner.rb
@@ -96,7 +96,7 @@ module RuboCop

       def begin_investigation(processed_source, offset:, original:)
         @cops.each do |cop|
-          cop.send(:begin_investigation, processed_source, offset: offset, original: original)
+          cop.begin_investigation(processed_source, offset: offset, original: original)
         end
       end
```

Reply:
This is required as part of this feature and not a change I want to see independently in maintenance. Can you squash your commits into one?